tools/gcc-bridge/compiler/src/main/java/org/renjin/gcc/codegen/constructors/Constructors.java

```diff
@@ -0,0 +1,335 @@
+    }
+
+    if (stringBuilder.length() < 1000) {
+
+      // Store the concatenated string in the Class' string pool
+
+      mv.aconst(stringBuilder.toString());
+      mv.invokestatic(Type.getType(BytePtr.class), "stringArray",
+          Type.getMethodDescriptor(Type.getType(Ptr.class), Type.getType(String.class)));
+
+    } else {
+
+      // Store in an external resource
+
+      byte[] bytes = stringBuilder.toString().getBytes(Charsets.US_ASCII);
+      String resourceName = Hashing.md5().hashBytes(bytes).toString();
+
+      try {
+        resourceWriter.writeResource(resourceName, bytes);
+      } catch (IOException e) {
+        throw new InternalCompilerException("Failed to write resource", e);
+      }
+
+      mv.aconst(mv.getOwnerClass());
+      mv.aconst(resourceName);
+      mv.invokestatic(Type.getType(BytePtr.class), "stringArrayFromResource",
+          Type.getMethodDescriptor(Type.getType(Ptr.class), Type.getType(Class.class), Type.getType(String.class)));
+    }
+
+    // Save the array to a local variable
+
+    LocalVar var = mv.getLocalVarAllocator().reserve(Type.getType(Ptr.class));
+
+    mv.store(var.getIndex(), Type.getType(Ptr.class));
+
+    return new VArrayExpr(arrayType, new VPtrExpr(var));
+  };
+
+  static ConstructorInterface charArray = (mv, resourceWriter, expr) -> {
+
+    if (!isArrayWithType(expr.getType(), c -> isIntegerWithPrecision(c, 16))) {
+      return null;
+    }
+
+    StringBuilder s = new StringBuilder();
+    for (GimpleConstructor.Element element : expr.getElements()) {
+      GimpleIntegerConstant constant = (GimpleIntegerConstant) element.getValue();
+      s.appendCodePoint(constant.getNumberValue().intValue());
+    }
+
+    JExpr array = new JExpr() {
+
+      @Nonnull
+      @Override
+      public Type getType() {
+        return Type.getType(Ptr.class);
+      }
+
+      @Override
+      public void load(@Nonnull MethodGenerator mv) {
+        mv.aconst(s.toString());
+        mv.invokestatic(Type.getType(CharPtr.class), "int16Array",
+            Type.getMethodDescriptor(Type.getType(Ptr.class),
+                Type.getType(String.class)));
+
+      }
+    };
+
+    return new VArrayExpr((GimpleArrayType) expr.getType(), new VPtrExpr(array));
+  };
+
+  /* Convenience functions */
+
+  private static ByteBuffer constantArrayToBuffer(
+      GimpleConstructor value,
+      Predicate<GimpleType> isElementType,
+      BufferConverter converter,
+      int byteSize,
+      int threshold
+  ) {
+    boolean isArrayType = isArrayWithType(value.getType(), isElementType);
+    if (!isArrayType) {
+      return null;
+    }
+    List<GimpleConstructor.Element> elements = value.getElements();
+    if (elements.size() < threshold) { // only stream large arrays
+      return null;
+    }
+    ByteBuffer buffer = ByteBuffer.allocate(elements.size() * byteSize).order(ByteOrder.LITTLE_ENDIAN);
+    for (GimpleConstructor.Element element : elements) {
+      converter.append(buffer, element);
```

Comment:
The constructor is not guaranteed to contain a `GimpleConstructor.Element` for each array element. Sometimes GCC will emit 'jagged' arrays with uninitialized elements, mostly from Fortran code. See here:
https://github.com/bedatadriven/renjin/blob/5eb75d9d0345ec46fc46ffd43284e8c62f214355/tools/gcc-bridge/compiler/src/main/java/org/renjin/gcc/codegen/array/ArrayTypeStrategy.java#L179

Reply:
I see.  Would inserting 0 values for these values be acceptable?

Reply:
Yes that's correct. I've extracted the normalizeArrayElements() logic to GimpleConstructor so that it can be used outside of ArrayTypeStrategy.

Reply:
No worries, judging by your latest commits it looks like you've got it covered :)